### PR TITLE
js: add MakeFullWrapper to expose exported methods and struct fields.

### PR DESCRIPTION
**Pull request by JounQin**
rebase for #790

close #790

---

Currently the documentation for js.MakeWrapper is:

> "MakeWrapper creates a JavaScript object which has wrappers for the
exported methods of i. Use explicit getter and setter methods to expose
struct fields to JavaScript."

Where the value a struct value (or more
interestingly a pointer to a struct value) we can actually auto-generate
getters and setters for exported fields in the JavaScript world, rather
than requiring explicit getters and setters to be defined on the Go
side.

We do this via a new MakeFullWrapper method.

### CODE REVIEW

**Comment by nevkontakte** [comment]
Thanks for doing the rebase! Upon the first pass the change looks good, but it is a part of the codebase I'm not very familiar with. I'll need a day or two more to understand this change in more detail.

Please disregard the failed check, it appears to be a bug in the check itself, unrelated to this PR.

File: js/js.go [hunk 1]
```diff
@@ -153,6 +153,103 @@ func MakeWrapper(i interface{}) *Object {
...
 
+// MakeFullWrapper creates a JavaScript object which has wrappers for the exported
+// methods of i, and, where i is a (pointer to a) struct value, wrapped getters
+// and setters
+// (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty)
+// for the non-embedded exported fields of i. Values accessed via these methods
+// and getters are themsevles wrapped when accessed, but an important point to
+// note is that a new wrapped value is created on each access.
+func MakeFullWrapper(i interface{}) *Object {
+	v := InternalObject(i)
```
Comment (nevkontakte) [inline js/js.go:164]: readability: let's use a more descriptive name, for example `internalObj`.

File: js/js.go [hunk 2]
```diff
@@ -153,6 +153,103 @@ func MakeWrapper(i interface{}) *Object {
...
+	c := v.Get("constructor")
+
+	o := Global.Get("Object").New()
```
Comment (nevkontakte) [inline js/js.go:167]: readability: let's use a more descriptive name here too, for example `wrapperObj`.

File: js/js.go [hunk 3]
```diff
@@ -153,6 +153,103 @@ func MakeWrapper(i interface{}) *Object {
...
+		op := Global.Get("Object").New()
+		for _, f := range fns {
+			f(op)
+		}
+		Global.Get("Object").Call("defineProperty", o, k, op)
+	}
+
+	defineProperty("__internal_object__", func(op *Object) {
+		op.Set("value", v)
+	})
```
Comment (nevkontakte) [inline js/js.go:179]: This function can be simplified greatly using the fact that GopherJS converts Go maps into JavaScript objects. Something like this:

```go
defineProperty := func(k string, descriptor M) {
  Global.Get("Object").Call("defineProperty", o, k, descriptor)
}

defineProperty("__internal_object__", M{
  "value": v,
})
```

(Note that `M` is a type defined in the `js` package: https://pkg.go.dev/github.com/gopherjs/gopherjs/js#M).

File: js/js.go [hunk 4]
```diff
@@ -153,6 +153,103 @@ func MakeWrapper(i interface{}) *Object {
...
+
+	{
+		// caculate a sensible type string
+
+		// we don't want to import any packages in this package
+		// so we do some string operations by hand
```
Comment (nevkontakte) [inline js/js.go:185]: readability: Comments should be full sentences starting with a capital letter and ending with punctuation. Same below.

**Review by nevkontakte — CHANGES_REQUESTED** [review]
@JounQin sorry for the delay, I had a very busy week at work. I left a few minor comments to improve code readability. I also pushed a fix for the failed GitHub Action, if you rebase on the latest master, the check should pass.

File: js/js.go [hunk 5]
```diff
@@ -153,6 +153,103 @@ func MakeWrapper(i interface{}) *Object {
 	return o
 }
 
+// MakeFullWrapper creates a JavaScript object which has wrappers for the exported
+// methods of i, and, where i is a (pointer to a) struct value, wrapped getters
+// and setters
+// (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty)
+// for the non-embedded exported fields of i. Values accessed via these methods
```
[the remark this answers is not in the document]
Comment (nevkontakte) [reply js/js.go:160]: I think it's more of a technical limitation, although not impossible to overcome. You'd have to do a lot more introspection of the exported object, which is difficult to do without reflect.

**Comment by nevkontakte** [comment]
@JounQin I'd like to merge this pull request and make a GopherJS release with it. If you need help with addressing the comments I left, you could enable "Allow edits and access to secrets by maintainers" on this PR, and I could give it a go.

**Comment by JounQin** [comment]
@nevkontakte 

image

I've already enabled this feature when PR open, feel free to change.

**Comment by nevkontakte** [comment]
Strange, I get an error when I try to push commits to it:

```
$ git push JounQuin getters -f
ERROR: Permission to JounQin/gopherjs.git denied to nevkontakte.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

I'll try to figure out what I'm going wrong :)

**Comment by nevkontakte** [comment]
@JounQin  I figured out the issue. Your repository is a fork of https://github.com/myitcv/gopherjs, not this repository. Apparently, in GitHub's eyes this [disqualifies](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork) me from being able to edit this pull request contents:

> You can only make commits on pull request branches that:
>  - are opened in a repository that you have push access to and that were __created from a fork of that repository__.

Do you need my assistance to address the review comments? I could take your branch and make yet another PR from it if so.

**Comment by JounQin** [comment]
@nevkontakte Please help to review again.

**Review by nevkontakte — APPROVED** [review]

**Comment by nevkontakte** [comment]
@JounQin thanks for addressing the comments, and thanks for the contribution!

**Pull request merged**